arviz/kde_utils.py

```diff
@@ -0,0 +1,895 @@
+
+    if len(custom_lims) != 2:
+        raise AttributeError(
+            f"`len(custom_lims)` must be 2, not {len(custom_lims)}.")
+
+    any_bool = any(isinstance(i, bool) for i in custom_lims)
+    if any_bool:
+        raise TypeError(
+            "Elements of `custom_lims` must be numeric or None, not bool.")
+
+    if custom_lims[0] is None:
+        custom_lims[0] = x_min
+
+    if custom_lims[1] is None:
+        custom_lims[1] = x_max
+
+    all_numeric = all(isinstance(i, (int, float, np.integer, np.float)) for i in custom_lims)
+    if not all_numeric:
+        raise TypeError((
+            f"Elements of `custom_lims` must be numeric or None.\n"
+            f"At least one of them is not."
+        ))
+
+    if not custom_lims[0] < custom_lims[1]:
+        raise AttributeError(
+            f"`custom_lims[0]` must be smaller than `custom_lims[1]`.")
+
+    return custom_lims
+
+def get_grid(x_min, x_max, x_std, extend_fct, grid_len, custom_lims, extend=True,
+            bound_correction=False):
+    """
+    Computes the grid that bins the data used to estimate the density function
+
+    Parameters
+    ----------
+    x_min : float
+        Minimum value of the data
+    x_max: float
+        Maximum value of the data.
+    x_std: float
+        Standard deviation of the data.
+    extend_fct: bool
+        Indicates the factor by which `x_std` is multiplied
+        to extend the range of the data.
+    grid_len: int
+        Number of bins
+    custom_lims: tuple or list
+        Custom limits for the domain of the density estimation.
+        Must be numeric of length 2.
+    extend: bool, optional
+        Whether to extend the range of the data or not.
+        Default is True.
+    bound_correction: bool, optional
+        Whether the density estimations performs boundary correction or not.
+        This does not impacts directly in the output, but is used
+        to override `extend`.
+        Default is False.
+
+    Returns
+    -------
+    grid_len: int
+        Number of bins
+    grid_min: float
+        Minimum value of the grid
+    grid_max: float
+        Maximum value of the grid
+
+    """
+
+    # Set up number of bins.
+    if grid_len < 100:
+        grid_len = 100
+    grid_len = int(grid_len)
+
+    # Set up domain
+    # `custom_lims` overrides `extend`
+    # `bound_correction` overrides `extend`
+    if custom_lims is not None:
+        custom_lims = check_custom_lims(custom_lims, x_min, x_max)
+        grid_min = custom_lims[0]
+        grid_max = custom_lims[1]
+    elif extend and not bound_correction:
+        grid_extend = extend_fct * x_std
+        grid_min = x_min - grid_extend
+        grid_max = x_max + grid_extend
+    else:
+        grid_min = x_min
+        grid_max = x_max
+    return grid_min, grid_max, grid_len
+
+# KDE Functions --------------------------------------------------------------------
+
+def kde(x, circular=False, **kwargs):
+    """
+    1 dimensional density estimation.
```

Comment:
again not sure if numpydoc lint will complain, we generally start docstrings in the same line as `"""`